Source/EmbeddedBoundary/WarpXFaceExtensions.cpp

```diff
@@ -0,0 +1,1114 @@
+            }
+        }
+
+        amrex::Real denom = local_avail(0, 1) * S(i - 1, j, k) +
+            local_avail(2, 1) * S(i + 1, j, k) +
+            local_avail(1, 0) * S(i, j - 1, k) +
+            local_avail(1, 2) * S(i, j + 1, k) +
+            local_avail(0, 0) * S(i - 1, j - 1, k) +
+            local_avail(2, 0) * S(i + 1, j - 1, k) +
+            local_avail(0, 2) * S(i - 1, j + 1, k) +
+            local_avail(2, 2) * S(i + 1, j + 1, k);
+
+        bool neg_face = true;
+
+        while(denom >= S_ext and neg_face and denom > 0){
+            neg_face = false;
+            for (int i_n = -1; i_n < 2; i_n++) {
+                for (int j_n = -1; j_n < 2; j_n++) {
+                    if(local_avail(i_n + 1, j_n + 1)){
+                        amrex::Real patch = S_ext * S(i + i_n, j + j_n, k) / denom;
+                        if(S_red(i + i_n, j + j_n, k) - patch <= 0) {
+                            neg_face = true;
+                            local_avail(i_n + 1, j_n + 1) = false;
+                        }
+                    }
+                }
+            }
+
+            denom = local_avail(0, 1) * S(i - 1, j, k) +
+                local_avail(2, 1) * S(i + 1, j, k) +
+                local_avail(1, 0) * S(i, j - 1, k) +
+                local_avail(1, 2) * S(i, j + 1, k) +
+                local_avail(0, 0) * S(i - 1, j - 1, k) +
+                local_avail(2, 0) * S(i + 1, j - 1, k) +
+                local_avail(0, 2) * S(i - 1, j + 1, k) +
+                local_avail(2, 2) * S(i + 1, j + 1, k);
+        }
+
+    }
+
+    for(int ii = 0; ii < 3; ii++) {
+        for (int jj = 0; jj < 3; jj++) {
+            n_borrow += local_avail(ii, jj);
+        }
+    }
+
+    return n_borrow;
+}
+
+/**
+* \brief Do the one-way extension
+*/
+amrex::Array1D<int, 0, 2>
+WarpX::ComputeOneWayExtensions() {
+#ifdef AMREX_USE_EB
+    auto const eb_fact = fieldEBFactory(maxLevel());
+    auto const &cell_size = CellSize(maxLevel());
+
+    int nelems_x;
+    int nelems_y;
+    int nelems_z;
+
+    // Do the extensions in the x-plane
+    int idim = 0;
+
+    for (amrex::MFIter mfi(*Bfield_fp[maxLevel()][idim]); mfi.isValid(); ++mfi) {
+
+        amrex::Box const &box = mfi.validbox();
+
+        auto const &Sx = m_face_areas[maxLevel()][idim]->array(mfi);
+        auto const &flag_ext_face_x = m_flag_ext_face[maxLevel()][idim]->array(mfi);
+        auto const &flag_info_face_x = m_flag_info_face[maxLevel()][idim]->array(mfi);
+        auto &borrowing_x = (*m_borrowing[maxLevel()][idim])[mfi];
+        auto const &borrowing_x_inds_pointer = borrowing_x.inds_pointer.array();
+        auto const &borrowing_x_size = borrowing_x.size.array();
+        amrex::Long ncells = box.numPts();
+        int* borrowing_x_inds = borrowing_x.inds.data();
+        uint8_t* borrowing_x_neigh_faces = borrowing_x.neigh_faces.data();
+        double* borrowing_x_area = borrowing_x.area.data();
```

Comment:
Does this have to be `double` instead of `amrex::Real` for a specific reason? Same question for similar `double` variables below (e.g. `borrowing_y_area`, `borrowing_z_area`, here and in other functions).

Reply:
There is no good reason for this. I changed them to `amrex::Real`